### PR TITLE
replaced MS_SHOULD_BE_FINAL general spotbugs exclude with individual excludes

### PR DESCRIPTION
**Pull request by StefanSpieker**
Replaced `MS_SHOULD_BE_FINAL` general spotbugs exclude with individual excludes. Added some comments for clarity. And specified one TODO in restricted since.

In case it is not useful to exclude at any point, comments are appreciated.

### Proposed changelog entries

* Internal: Replaced general Spotbugs exclude for `MS_SHOULD_BE_FINAL` with individual supressions
* Introduced new systemproperty `hudson.triggers.SlowTriggerAdminMonitor.maxEntries` to define maximum entries.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/tasks/_maven/Maven3MojoNote.java [hunk 1]
```diff
@@ -65,5 +66,6 @@ public String getDisplayName() {
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/_maven/Maven3MojoNote.java:69]: Should it actually be final? If no, some justification would be cool

File: core/src/main/java/hudson/tasks/_maven/MavenErrorNote.java [hunk 2]
```diff
@@ -52,6 +53,7 @@ public String getDisplayName() {
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/_maven/MavenErrorNote.java:56]: Same?

File: core/src/main/java/hudson/tasks/_maven/MavenMojoNote.java [hunk 3]
```diff
@@ -57,5 +58,6 @@ public String getDisplayName() {
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/_maven/MavenMojoNote.java:61]: Same?

File: core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java [hunk 4]
```diff
@@ -28,6 +29,7 @@
     @NonNull
     private final Map<String, Value> errors = new ConcurrentHashMap<>();
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static int MAX_ENTRIES = 10;
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java:33]: Maybe makes sense to add a system property support. Not a blocker

**Review by oleg-nenashev — COMMENTED** [review]
IMHO some fields should really be final. Not blocking though, this warning is rather a noise than help at the moment

File: core/src/main/java/hudson/tasks/_maven/Maven3MojoNote.java (hunk 1 above)
Comment (StefanSpieker) [reply core/src/main/java/hudson/tasks/_maven/Maven3MojoNote.java:69]: I guess noone will try to modify the pattern with script console, so I'm setting it final.

File: core/src/main/java/hudson/tasks/_maven/MavenErrorNote.java (hunk 2 above)
Comment (StefanSpieker) [reply core/src/main/java/hudson/tasks/_maven/MavenErrorNote.java]: I guess noone will try to modify the pattern with script console, so I'm setting it final.

File: core/src/main/java/hudson/tasks/_maven/MavenMojoNote.java (hunk 3 above)
Comment (StefanSpieker) [reply core/src/main/java/hudson/tasks/_maven/MavenMojoNote.java]: I guess noone will try to modify the pattern with script console, so I'm setting it final.

File: core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java (hunk 4 above)
Comment (StefanSpieker) [reply core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java]: Done. I guess adding this to the documentation would make sense ...

File: core/src/main/java/hudson/tasks/_maven/MavenWarningNote.java [hunk 5]
```diff
@@ -54,5 +55,6 @@ public String getDisplayName() {
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static Pattern PATTERN = Pattern.compile("^\\[WARNING\\]");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/tasks/_maven/MavenWarningNote.java:59]: Taking the previous changes, should be final as well to keep it consistent

**Review by oleg-nenashev — APPROVED** [review]

**Comment by timja** [comment]
conflicts fyi

File: core/src/main/java/hudson/tasks/_maven/MavenWarningNote.java (hunk 5 above)
Comment (StefanSpieker) [reply core/src/main/java/hudson/tasks/_maven/MavenWarningNote.java]: You are right. added final and removed suppression

**Comment by StefanSpieker** [comment]
> 
> 
> conflicts fyi

resolved.

**Review by oleg-nenashev — APPROVED** [review]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

File: core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java [hunk 6]
```diff
@@ -28,7 +30,8 @@
     @NonNull
     private final Map<String, Value> errors = new ConcurrentHashMap<>();
 
-    public static int MAX_ENTRIES = 10;
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
+    public static /* Script Console modifiable */ int MAX_ENTRIES = SystemProperties.getInteger(SlowTriggerAdminMonitor.class.getName() + ".maxEntries", 10);
```
Comment (daniel-beck) [inline core/src/main/java/hudson/triggers/SlowTriggerAdminMonitor.java:34]: Needs corresponding update to https://www.jenkins.io/doc/book/managing/system-properties/ and perhaps even separate changelog entry.

File: core/src/main/java/jenkins/tools/GlobalToolConfiguration.java [hunk 7]
```diff
@@ -107,6 +108,7 @@ private boolean configureDescriptor(StaplerRequest req, JSONObject json, Descrip
         return d.configure(req, js);
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static Predicate<Descriptor> FILTER = input -> input.getCategory() instanceof ToolConfigurationCategory;
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/tools/GlobalToolConfiguration.java:112]: This looks like something that should be final.

File: core/src/main/java/jenkins/util/xstream/XStreamDOM.java [hunk 8]
```diff
@@ -537,5 +538,6 @@ public XStreamDOM unmarshalElement(HierarchicalStreamReader r, UnmarshallingCont
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static XmlFriendlyReplacer REPLACER = new XmlFriendlyReplacer();
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/util/xstream/XStreamDOM.java:542]: Potentially also better off final?

File: core/src/main/java/jenkins/model/IdStrategy.java [hunk 9]
```diff
@@ -55,6 +56,7 @@
     /**
      * The default case insensitive strategy.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static IdStrategy CASE_INSENSITIVE = new CaseInsensitive();
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/IdStrategy.java:60]: I am surprised this isn't final.

File: core/src/main/java/jenkins/install/SetupWizard.java [hunk 10]
```diff
@@ -94,6 +94,7 @@ public SetupWizard() {
     /**
      * The security token parameter name
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static String initialSetupAdminUserName = "admin";
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/install/SetupWizard.java:98]: This looks like something that should be final.

File: core/src/main/java/hudson/util/io/ArchiverFactory.java [hunk 11]
```diff
@@ -57,6 +59,7 @@
     /**
      * Zip format.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static ArchiverFactory ZIP = new ZipArchiverFactory();
```
Comment (daniel-beck) [inline core/src/main/java/hudson/util/io/ArchiverFactory.java:63]: This looks like something that should be final.

File: core/src/main/java/hudson/util/io/ArchiverFactory.java [hunk 12]
```diff
@@ -47,6 +48,7 @@
     /**
      * Uncompressed tar format.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static ArchiverFactory TAR = new TarArchiverFactory(TarCompression.NONE);
```
Comment (daniel-beck) [inline core/src/main/java/hudson/util/io/ArchiverFactory.java:52]: This looks like something that should be final.

File: core/src/main/java/hudson/util/ColorPalette.java [hunk 13]
```diff
@@ -46,6 +47,7 @@
     /**
      * Color list usable for generating line charts.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static List<Color> LINE_GRAPH = Collections.unmodifiableList(Arrays.asList(
```
Comment (daniel-beck) [inline core/src/main/java/hudson/util/ColorPalette.java:51]: This looks like something that should be final.

File: core/src/main/java/hudson/util/ChartUtil.java [hunk 14]
```diff
@@ -253,6 +253,7 @@ public static void adjustChebyshev(CategoryDataset dataset, NumberAxis yAxis) {
         yAxis.setRange(min,max);
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static double CHEBYSHEV_N = 3;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/util/ChartUtil.java:257]: This looks like something that should be final.

File: core/src/main/java/hudson/triggers/Trigger.java [hunk 15]
```diff
@@ -301,6 +302,7 @@ public void run(AbstractProject p) {
         }
     }
 
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static long CRON_THRESHOLD = 1000*30;    // Default threshold 30s
```
Comment (daniel-beck) [inline core/src/main/java/hudson/triggers/Trigger.java:306]: Probably could use a system property as well.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 16]
```diff
@@ -5326,6 +5328,7 @@ public boolean shouldShowStackTrace() {
     /**
      * Automatically try to launch an agent when Jenkins is initialized or a new agent computer is created.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static boolean AUTOMATIC_SLAVE_LAUNCH = true;
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:5332]: Probably could use a system property as well.

File: core/src/main/java/hudson/security/GlobalSecurityConfiguration.java [hunk 17]
```diff
@@ -194,6 +195,7 @@ public Permission getRequiredPermission() {
 
     @Restricted(NoExternalUse.class)
     @RestrictedSince("2.222")
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static Predicate<Descriptor> FILTER = input -> input.getCategory() instanceof GlobalConfigurationCategory.Security;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/GlobalSecurityConfiguration.java:199]: This looks like it should be final.

File: core/src/main/java/hudson/scm/SCMRevisionState.java [hunk 18]
```diff
@@ -44,7 +45,7 @@ public String getUrlName() {
       So instead, here we opt to a design where we tell SCM upfront about what we are comparing
       against (baseline), and have it give us the new state and degree of change in PollingResult.
      */
-
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static SCMRevisionState NONE = new None();
```
Comment (daniel-beck) [inline core/src/main/java/hudson/scm/SCMRevisionState.java:49]: This looks like it should be final.

File: core/src/main/java/hudson/diagnosis/HudsonHomeDiskUsageChecker.java [hunk 19]
```diff
@@ -66,6 +67,7 @@ protected void doRun() {
     /**
      * Gets the minimum amount of space to check for, with a default of 10GB
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static long FREE_SPACE_THRESHOLD = Long.getLong(
```
Comment (daniel-beck) [inline core/src/main/java/hudson/diagnosis/HudsonHomeDiskUsageChecker.java:71]: This should access `SystemProperties` for the value (unless that doesn't support longs?)

File: core/src/main/java/hudson/Proc.java [hunk 20]
```diff
@@ -505,6 +505,7 @@ public OutputStream getStdin() {
     /**
      * Debug switch to have the thread display the process it's waiting for.
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static boolean SHOW_PID = false;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Proc.java:509]: Probably could use a system property, unsure whether it's something you want on startup.

File: core/src/main/java/hudson/PluginManager.java [hunk 21]
```diff
@@ -2430,5 +2431,6 @@ public boolean hasAdoptThisPluginLabel(PluginWrapper plugin) {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(PluginManager.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginManager.java:2435]: Should use `SystemProperties`.

File: core/src/main/java/hudson/Plugin.java [hunk 22]
```diff
@@ -309,6 +311,7 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(Plugin.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Plugin.java:315]: Should use `SystemProperties`.

File: core/src/main/java/hudson/FilePath.java [hunk 23]
```diff
@@ -2906,6 +2907,7 @@ public String validateAntFileMask(final String fileMasks, final int bound) throw
      * Default bound for {@link #validateAntFileMask(String, int, boolean)}.
      * @since 1.592
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static int VALIDATE_ANT_FILE_MASK_BOUND = Integer.getInteger(FilePath.class.getName() + ".VALIDATE_ANT_FILE_MASK_BOUND", 10000);
```
Comment (daniel-beck) [inline core/src/main/java/hudson/FilePath.java:2911]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/AbstractItem.java [hunk 24]
```diff
@@ -957,6 +958,7 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(AbstractItem.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/AbstractItem.java:962]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/Computer.java [hunk 25]
```diff
@@ -1601,6 +1601,7 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(Computer.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Computer.java:1605]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/FileParameterValue.java [hunk 26]
```diff
@@ -72,6 +73,7 @@
      * It's not recommended to enable for security reasons. That option is only present for backward compatibility.
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean ALLOW_FOLDER_TRAVERSAL_OUTSIDE_WORKSPACE = 
             Boolean.getBoolean(FileParameterValue.class.getName() + ".allowFolderTraversalOutsideWorkspace");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/FileParameterValue.java:78]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/Run.java [hunk 27]
```diff
@@ -2659,6 +2660,7 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(Run.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Run.java:2664]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 28]
```diff
@@ -2628,6 +2629,7 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(UpdateCenter.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateCenter.java:2633]: Should use `SystemProperties`.

File: core/src/main/java/hudson/model/User.java [hunk 29]
```diff
@@ -127,6 +128,7 @@
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(User.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/User.java:132]: Should use `SystemProperties`.

File: core/src/main/java/hudson/search/Search.java [hunk 30]
```diff
@@ -415,6 +417,7 @@ public Object getTarget() {
     /**
      * Escape hatch for StaplerProxy-based access control
      */
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     @Restricted(NoExternalUse.class)
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(Search.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/search/Search.java:422]: Should use `SystemProperties`.

File: core/src/main/java/jenkins/model/StandardArtifactManager.java [hunk 31]
```diff
@@ -50,6 +51,7 @@
 
     @Restricted(NoExternalUse.class)
     @VisibleForTesting
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static FilePath.TarCompression TAR_COMPRESSION = Boolean.getBoolean(StandardArtifactManager.class.getName() + ".disableTrafficCompression")
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/StandardArtifactManager.java:55]: Should use `SystemProperties`.

File: core/src/main/java/hudson/logging/LogRecorderManager.java [hunk 32]
```diff
@@ -219,5 +220,6 @@ public Object getTarget() {
      * Escape hatch for StaplerProxy-based access control
      */
     @Restricted(NoExternalUse.class)
+    @SuppressFBWarnings("MS_SHOULD_BE_FINAL")
     public static /* Script Console modifiable */ boolean SKIP_PERMISSION_CHECK = Boolean.getBoolean(LogRecorderManager.class.getName() + ".skipPermissionCheck");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/logging/LogRecorderManager.java:224]: Should use `SystemProperties`.

**Review by daniel-beck — APPROVED** [review]
Looks reasonable for what it does. Some suggestions for further improvements while we're here, but could also be done at a later time.

Have not done more than superficial compatibility checks for things I suggest newly be made `final`.

**Comment by StefanSpieker** [comment]
Thanks for the suggestions @daniel-beck 
I will take a look if I find some time during the week.

**Pull request merged**